cmd/nsdns/watch.go

```diff
@@ -52,37 +57,59 @@ func watchCommand() *cobra.Command {
 				return err
 			}
 
-			informerFactory := informers.NewSharedInformerFactory(clientset, time.Minute)
+			var stop chan<- struct{}
+			informerFactory := informer(dm, clientset)
 
-			informerFactory.Networking().V1().Ingresses().Informer().AddEventHandler(
-				cache.ResourceEventHandlerFuncs{
-					AddFunc: func(obj interface{}) {
-						ingress := obj.(*networkingv1.Ingress)
+			if leaseName != "" {
+				// Validate fields present in env
+				podNamespace := os.Getenv("POD_NAMESPACE")
+				if podNamespace == "" {
+					return errors.New("must provide POD_NAMESPACE to use leases")
+				}
 
-						if err := dm.HandleIngressExists(ingress); err != nil {
-							log.Error(err)
-						}
-					},
-					DeleteFunc: func(obj interface{}) {
-						ingress := obj.(*networkingv1.Ingress)
+				podName := os.Getenv("POD_NAME")
+				if podName == "" {
+					return errors.New("must provide POD_NAME to use leases")
+				}
 
-						if err := dm.HandleIngressDeleted(ingress); err != nil {
-							log.Error(err)
-						}
+				leConfig := leaderelection.LeaderElectionConfig{
+					Lock: &resourcelock.LeaseLock{
+						LeaseMeta: metav1.ObjectMeta{
+							Name:      leaseName,
+							Namespace: podNamespace,
+						},
+						Client: clientset.CoordinationV1(),
+						LockConfig: resourcelock.ResourceLockConfig{
+							Identity: podName,
+						},
 					},
-					UpdateFunc: func(old, new interface{}) {
-						ingress := new.(*networkingv1.Ingress)
-
-						if err := dm.HandleIngressExists(ingress); err != nil {
-							log.Error(err)
-						}
+					ReleaseOnCancel: true,
+					LeaseDuration:   30 * time.Second,
+					RenewDeadline:   15 * time.Second,
+					RetryPeriod:     5 * time.Second,
+					Callbacks: leaderelection.LeaderCallbacks{
+						OnStartedLeading: func(ctx context.Context) {
+							if err := dm.UpdateCache(); err != nil {
```

Comment:
Doesn't _really_ work, for a couple reasons.
1. When the leader fails, there's a ~30 second window where the pod can just try to come back up, and take back its own lease, and then immediately fail again because the reason it failed it still live.
2. If that doesn't happen, the new leader that gets elected will probably hit this block anyways, since the main reason this process exits is because of `dm.UpdateCache()` failing due to Namesilo not being accessible. Records could have gone stale on the one node, and then... this assumption is broken.